Improve apply retry

kramerul opened the pull request:
This PR is mostly related to issues #82 and #87.

* `util.Retry` was replaced by `util.RetryOnError` which is a copy of `k8s.io/client-go/util/retry.OnError`. It seems to be the standard way to do retries with newer versions of the k8s client library.
* The two errors `http2.GoAwayError` and `errors.StatusError` with `err.ErrStatus.Reason == metav1.StatusReasonServiceUnavailable` are automatically retried.
* List functions are also retried now

kramerul committed:
Improve apply retry. Relates #82 #87

pivotal-issuemaster commented:
@kramerul Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/pivotal?repositoryId=k14s/kapp&pullRequestId=122)!

[Click here](https://cla.pivotal.io/sync/pivotal?repositoryId=k14s/kapp&pullRequestId=122) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

pivotal-issuemaster commented:
@kramerul Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/pivotal?repositoryId=k14s/kapp&pullRequestId=122)!

cppforlife commented:
few comments:

IsRetryableErr was meant to represent error that could be retried at a higher level due to resource conflicts (ie one resource waiting for the other). i think we should keep it that way. http2 errors should be retried at a lower level and not leak out to higher level retry mechanisms.

i would also like to stick with util.Retry to keep things very simple for now.

cppforlife commented:
IsRetryableErr should probably be renamed to something like IsResourceChangeBlockedErr to indicate better that resource change is blocked on somethign else in the system (like another resource being changed).

kramerul committed:
Improve apply retry. Add Dimitry's feedback.

kramerul commented:
Hi @cppforlife,

I renamed `IsRetryableErr` to `IsResourceChangeBlockedErr` and reverted the usage of `RetryOnError` although in my point of view the code looked simpler and an exponential backoff is missing now.
